File: fixtures/coding/batch-ingestion-go/transform_test.go
package pipeline

import (
	"reflect"
	"testing"
	"time"
)

func fixedRecord(payload Payload, recordType ...string) Record {
	rt := "test"
	if len(recordType) > 0 {
		rt = recordType[0]
	}
	return Record{RecordType: rt, TS: time.Date(2025, 1, 1, 0, 0, 0, 0, time.UTC), Payload: payload, Tags: StringSet{}}
}

func TestNormalizePayloadSnakeCaseConversion(t *testing.T) {
	result := NormalizePayload(fixedRecord(Payload{"UserName": "alice", "LastLoginCount": 42}))
	if _, ok := result.Payload["user_name"]; !ok {
		t.Fatalf("missing user_name: %#v", result.Payload)
	}
	if _, ok := result.Payload["last_login_count"]; !ok {
		t.Fatalf("missing last_login_count: %#v", result.Payload)
	}
	if _, ok := result.Payload["UserName"]; ok {
		t.Fatalf("old key still present: %#v", result.Payload)
	}
}

func TestNormalizePayloadTrimsStrings(t *testing.T) {
	result := NormalizePayload(fixedRecord(Payload{"name": "  alice  ", "num": 42}))
	if result.Payload["name"] != "alice" || result.Payload["num"] != 42 {
		t.Fatalf("unexpected payload: %#v", result.Payload)
	}
}

func TestNormalizePayloadDoesNotMutateInput(t *testing.T) {
	r := fixedRecord(Payload{"OriginalKey": 1})
	original := Payload{"OriginalKey": 1}
	NormalizePayload(r)
	if !reflect.DeepEqual(r.Payload, original) {
		t.Fatalf("input mutated: %#v", r.Payload)
	}
}

func TestNormalizePayloadEmptyPayload(t *testing.T) {
	result := NormalizePayload(fixedRecord(Payload{}))
	if len(result.Payload) != 0 {
		t.Fatalf("expected empty payload, got %#v", result.Payload)
	}
}

func TestNormalizePayloadAlreadySnakeCase(t *testing.T) {
	result := NormalizePayload(fixedRecord(Payload{"already_snake": 1, "also_ok": "val"}))
	if _, ok := result.Payload["already_snake"]; !ok {
		t.Fatalf("missing already_snake")
	}
	if _, ok := result.Payload["also_ok"]; !ok {
		t.Fatalf("missing also_ok")
	}
}

func TestFlattenPayloadOneLevel(t *testing.T) {
	result := FlattenPayload(fixedRecord(Payload{"user": map[string]any{"name": "alice", "id": 7}, "page": "/home"}))
	if result.Payload["user_name"] != "alice" || result.Payload["user_id"] != 7 || result.Payload["page"] != "/home" {
		t.Fatalf("unexpected flattened payload: %#v", result.Payload)
	}
	if _, ok := result.Payload["user"]; ok {
		t.Fatalf("old nested key still present: %#v", result.Payload)
	}
}

func TestFlattenPayloadNonMapValuesNotFlattened(t *testing.T) {
	result := FlattenPayload(fixedRecord(Payload{"count": 5, "active": true}))
	if result.Payload["count"] != 5 || result.Payload["active"] != true {
		t.Fatalf("unexpected payload: %#v", result.Payload)
	}
}

func TestFlattenPayloadOnlyOneLevel(t *testing.T) {
	result := FlattenPayload(fixedRecord(Payload{"nested": map[string]any{"inner": map[string]any{"deep": "value"}, "id": 1}}))
	inner, ok := result.Payload["nested_inner"].(map[string]any)
	if !ok || inner["deep"] != "value" {
		t.Fatalf("inner should remain nested, got %#v", result.Payload["nested_inner"])
	}
}

func TestFlattenPayloadDoesNotMutateInput(t *testing.T) {
	r := fixedRecord(Payload{"a": map[string]any{"b": 1}, "c": 2})
	original := Payload{"a": map[string]any{"b": 1}, "c": 2}
	FlattenPayload(r)
	if !reflect.DeepEqual(r.Payload, original) {
		t.Fatalf("input mutated: %#v", r.Payload)
	}
}

func TestTransformBatchPipeline(t *testing.T) {
	records := []Record{
		fixedRecord(Payload{"User": map[string]any{"Name": "alice"}, "Score": 100}, "a"),
		fixedRecord(Payload{"Item": "book"}, "b"),
	}
	results := TransformBatch(records)
	if len(results) != 2 || results[0].Payload["user_name"] != "alice" || results[0].Payload["score"] != 100 {
		t.Fatalf("unexpected transform results: %#v", results)
	}
}

func TestTransformBatchPreservesOrder(t *testing.T) {
	results := TransformBatch([]Record{fixedRecord(Payload{"A": 1}, "z"), fixedRecord(Payload{"B": 2}, "a")})
	if len(results) != 2 || results[0].RecordType != "z" || results[1].RecordType != "a" {
		t.Fatalf("order not preserved: %#v", results)
	}
}

func TestTransformBatchDoesNotMutateInput(t *testing.T) {
	records := []Record{fixedRecord(Payload{"Key": 1}, "x")}
	TransformBatch(records)
	if _, ok := records[0].Payload["Key"]; !ok || len(records[0].Payload) != 1 {
		t.Fatalf("input mutated: %#v", records[0].Payload)
	}
}

File: fixtures/coding/batch-ingestion-ts/tests/transform.test.ts
import { describe, expect, test } from "vitest";
import type { IngestRecord } from "../src/types";
import { normalizePayload, flattenPayload, transformBatch } from "../src/transform";

const ts = new Date("2025-01-01T00:00:00Z");
const record = (payload: Record<string, unknown>, recordType = "test"): IngestRecord => ({
  recordType,
  ts,
  payload,
  tags: new Set<string>(),
});

describe("normalizePayload", () => {
  test("converts PascalCase/CamelCase keys to snake_case", () => {
    const result = normalizePayload(record({ UserName: "alice", LastLoginCount: 42 }));
    expect(Object.keys(result.payload)).toContain("user_name");
    expect(Object.keys(result.payload)).toContain("last_login_count");
    expect(Object.keys(result.payload)).not.toContain("UserName");
  });

  test("trims string values", () => {
    const result = normalizePayload(record({ name: "  alice  ", num: 42 }));
    expect(result.payload.name).toBe("alice");
    expect(result.payload.num).toBe(42);
  });

  test("does not mutate input payload", () => {
    const r = record({ OriginalKey: 1 });
    const original = { ...r.payload };
    normalizePayload(r);
    expect(r.payload).toEqual(original);
  });

  test("empty payload remains empty", () => {
    expect(normalizePayload(record({})).payload).toEqual({});
  });

  test("already snake_case keys stay present", () => {
    const result = normalizePayload(record({ already_snake: 1, also_ok: "val" }));
    expect(result.payload).toHaveProperty("already_snake");
    expect(result.payload).toHaveProperty("also_ok");
  });
});

describe("flattenPayload", () => {
  test("flattens one level", () => {
    const result = flattenPayload(record({ user: { name: "alice", id: 7 }, page: "/home" }));
    expect(result.payload.user_name).toBe("alice");
    expect(result.payload.user_id).toBe(7);
    expect(result.payload.page).toBe("/home");
    expect(result.payload).not.toHaveProperty("user");
  });

  test("non-object values are not flattened", () => {
    const result = flattenPayload(record({ count: 5, active: true }));
    expect(result.payload.count).toBe(5);
    expect(result.payload.active).toBe(true);
  });

  test("only flattens one level", () => {
    const result = flattenPayload(record({ nested: { inner: { deep: "value" }, id: 1 } }));
    expect(result.payload).toHaveProperty("nested_inner");
    expect(result.payload.nested_inner).toEqual({ deep: "value" });
  });

  test("arrays and dates are not flattened as plain objects", () => {
    const when = new Date("2025-01-02T00:00:00Z");
    const result = flattenPayload(record({ items: [1, 2], when }));
    expect(result.payload.items).toEqual([1, 2]);
    expect(result.payload.when).toBe(when);
  });

  test("does not mutate input payload", () => {
    const r = record({ a: { b: 1 }, c: 2 });
    const original = { ...r.payload };
    flattenPayload(r);
    expect(r.payload).toEqual(original);
  });
});

describe("transformBatch", () => {
  test("normalizes then flattens every record", () => {
    const records = [
      record({ User: { Name: "alice" }, Score: 100 }, "a"),
      record({ Item: "book" }, "b"),
    ];
    const results = transformBatch(records);
    expect(results).toHaveLength(2);
    expect(results[0]?.payload).toHaveProperty("user_name");
    expect(results[0]?.payload.score).toBe(100);
  });

  test("preserves order", () => {
    const results = transformBatch([record({ A: 1 }, "z"), record({ B: 2 }, "a")]);
    expect(results.map((r) => r.recordType)).toEqual(["z", "a"]);
  });

  test("does not mutate input records", () => {
    const records = [record({ Key: 1 }, "x")];
    transformBatch(records);
    expect(Object.keys(records[0]?.payload ?? {})).toEqual(["Key"]);
  });
});
